R/zchunk_LA115.RooftopPV.R

```diff
@@ -30,43 +28,125 @@ module_gcam.usa_LA115.RooftopPV_DISABLED <- function(command, ...) {
+      filter(NREL_Com_PV_supply_curve, Rel_Cost == min(Rel_Cost))$MW /
+      HOURS_PER_YEAR -> max_comm_capacity_factor
+
+    # Get minimum levelized electricity cost (LEC) for residential PV in 2005USD / KW
+    PV_resid_min_LEC_2005 <- (PV.RESID_INSTALLED_COST / PV.DERATING_FACTOR) * PV_FCR /
+      (max_resid_capacity_factor * HOURS_PER_YEAR) +
+      PV.RESID_OM / (max_resid_capacity_factor * HOURS_PER_YEAR)
+    # ... and in 1975USD / GJ
+    PV_resid_min_LEC_1975 <- PV_resid_min_LEC_2005 * gdp_deflator(1975, 2005) / CONV_KWH_GJ
+
+    # Get minimum levelized electricity cost (LEC) for commercial PV in 2005USD / KW
+    PV_comm_min_LEC_2005 <- (PV.COMM_INSTALLED_COST / PV.DERATING_FACTOR ) * PV_FCR /
+      (max_comm_capacity_factor * HOURS_PER_YEAR) +
+      PV.COMM_OM / (max_comm_capacity_factor * HOURS_PER_YEAR)
+    # ... and in 1975USD / GJ
+    PV_comm_min_LEC_1975 <- PV_comm_min_LEC_2005 * gdp_deflator(1975, 2005) / CONV_KWH_GJ
+
+    # COMPUTE SUPPLY CURVES
+
+    # Get residential supply curve
+    NREL_Res_PV_supply_curve %>%
+      mutate(p = Relative_Cost * PV_resid_min_LEC_1975,
+             generation = MWh / CONV_BIL_MIL) %>%  # << convert to GWh (like commercial)
+      rename(state_name = State) %>%
+      select(state_name, p, generation) ->
+      L115.pv_sc_res
+
+    # Get commercial supply curve
+    NREL_Com_PV_supply_curve %>%
+      mutate(p = Rel_Cost * PV_comm_min_LEC_1975,
+             generation = GWH) %>%
+      rename(state_name = State) %>%
+      select(state_name, p, generation) ->
+      L115.pv_sc_com
+
+    # Combine supply curves, add state abbrv., and rebase p
+    bind_rows(L115.pv_sc_res, L115.pv_sc_com) %>%
+      left_join_error_no_match(select(states_subregions, state_name, state),
+                               by = "state_name") %>%
+      arrange(state_name) %>%
+      select(-state_name, state, p, generation) %>%
+      mutate(p = p - min(p)) %>%
+      arrange(p) ->
+      # ^^ note: ordering of generation for matching p values will affect supply curve parameters (fitted below).
+      L115.pv_sc
+
+    unique(select(L115.pv_sc, state)) %>% mutate(p = 0, generation = 0) %>%
+      # ^^ adds a 0,0 row to all states for supply curve fitting)
+      bind_rows(L115.pv_sc) %>%
+      arrange(state, p) %>%
+      # ^^ sort for cumulative column calcs to follow
+      group_by(state) %>% mutate(cumul = cumsum(generation),
+                                 percent_cumul = cumul / sum(generation)) ->
+      # ^^ computes cumulative sum (abs. and percent) for each state separately
+      L115.pv_sc_State
+
+    # Get mid-price for each state
+    L115.pv_sc_State %>%
+      filter(percent_cumul > 0.5) %>%
+      # mid-price taken as the first price after the 50th percentile
+      summarise(mid_p = first(p)) -> L115.pv_midPrice
+
+    # Create smoothing functions
+    smooth_res_curve_approx <- function( b, midp, p ) {
```

Comment:
Are these used anywhere else? I'm wondering if their definitions belong in `pipeline_helpers.R` or `module_helpers.R`

Reply:
Don't seem to be elsewhere (not sure though). I'd be for keeping them in here so that anyone reading the script can easily see the function being fitted.